File: internal/domain/usecase/user_usecase/get_user_usecase.go
package user_usecase

import (
	"blueberry_homework/dto/response"
	"errors"

	"github.com/gocql/gocql"
)

// GetUser는 ID로 특정 사용자 정보를 가져옵니다.
func (u *UserUsecase) GetUser(id string) (response.UserResponse, error) {
	// 유저 ID UUID 변환
	parsedId, err := gocql.ParseUUID(id)
	if err != nil {
		return response.UserResponse{}, err
	}

	userExist, err := u.repo.FindById(parsedId)
	if err != nil {
		return response.UserResponse{}, err
	}
	if !userExist {
		return response.UserResponse{}, errors.New("user not found")
	}

	return u.repo.GetUser(parsedId)
}

File: src/internal/domain/usecases/user_usecases/get_user_usecase.rs
use std::sync::Arc;

use uuid::Uuid;

use crate::{
    dto::res::user_response::UserResponse,
    internal::domain::repository_interface::user_repository::UserRepository
};

pub struct GetUserUsecase {
    user_repo: Arc<dyn UserRepository + Send + Sync>,
}

impl GetUserUsecase {
    pub fn new(user_repo: Arc<dyn UserRepository + Send + Sync>) -> Self {
        Self { user_repo }
    }

    pub async fn get_user_usecase(&self, user_id: String) -> Result<UserResponse, String> {
        // input id is string type, so need to parse it to uuid type
        let parsed_user_id =
            Uuid::parse_str(&user_id)
                .map_err(|e| format!("Invalid user id: {}", e))?;

        let check_user = self.user_repo.find_by_id(parsed_user_id).await?;
        if !check_user {
            return Err(format!("User not found"));
        }

        let user = self.user_repo.get_user(parsed_user_id).await?;

        Ok(user)
    }
}
